### ISSUE TITLE
p-adic fields timing out

### ISSUE THREAD

**Issue by roed314** [LMFDB/lmfdb#4884]
From the flasklog on prod:
```
psycopg2.errors.QueryCanceled: canceling statement due to statement timeout
 happens while executing b'SELECT "id" FROM "nf_fields" WHERE "local_algs" @> ARRAY[\'2.8.16.6\']::text[] LIMIT 1'
[2021-11-25 07:09:13 UTC] 500 error on URL https://www.lmfdb.org/padicField/2.8.16.6 ()
```
I'm not sure if this is a missing index or if we need to change how things are stored (if `nf_fields` is just too big).

**Comment by edgarcosta** [comment]
```
-------------------------------------------------------------------------------------------------------------------------
 Limit  (cost=0.00..35.88 rows=1 width=8) (actual time=4574.014..4574.018 rows=1 loops=1)
   ->  Seq Scan on nf_fields  (cost=0.00..1556618.55 rows=43387 width=8) (actual time=4574.010..4574.012 rows=1 loops=1)
         Filter: (local_algs @> '{2.8.16.6}'::text[])
         Rows Removed by Filter: 12313185
 Planning Time: 0.423 ms
 Execution Time: 4574.047 ms
(6 rows)
```
I don't get why it doesn't use the `nf_fields_local_algs_gin` index....

**Comment by edgarcosta** [comment]
ok, the reason is the buffered heaps, if one changes the local data one gets something more reasonable:
```
EXPLAIN (ANALYZE, COSTS, VERBOSE, BUFFERS) SELECT "id" FROM "nf_fields" WHERE "local_algs" @> ARRAY['2.8.16.7']::text[] LIMIT 1;
                                                                   QUERY PLAN
------------------------------------------------------------------------------------------------------------------------------------------------
 Limit  (cost=78.76..82.56 rows=1 width=8) (actual time=30.244..30.247 rows=1 loops=1)
   Output: id
   Buffers: shared hit=3 read=3
   ->  Bitmap Heap Scan on public.nf_fields  (cost=78.76..23000.41 rows=6034 width=8) (actual time=30.242..30.243 rows=1 loops=1)
         Output: id
         Recheck Cond: (nf_fields.local_algs @> '{2.8.16.7}'::text[])
         Heap Blocks: exact=1
         Buffers: shared hit=3 read=3
         ->  Bitmap Index Scan on nf_fields_local_algs_gin  (cost=0.00..77.26 rows=6034 width=0) (actual time=29.716..29.717 rows=3104 loops=1)
               Index Cond: (nf_fields.local_algs @> '{2.8.16.7}'::text[])
               Buffers: shared hit=3 read=2
 Planning:
   Buffers: shared hit=1
 Planning Time: 0.249 ms
 Execution Time: 30.275 ms
(15 rows)
```

and if one goes back to original query one gets:
```
EXPLAIN (ANALYZE, COSTS, VERBOSE, BUFFERS) SELECT "id" FROM "nf_fields" WHERE "local_algs" @> ARRAY['2.8.16.6']::text[] LIMIT 1;
                                                        QUERY PLAN
--------------------------------------------------------------------------------------------------------------------------
 Limit  (cost=0.00..42.98 rows=1 width=8) (actual time=0.023..0.024 rows=1 loops=1)
   Output: id
   Buffers: shared hit=1
   ->  Seq Scan on public.nf_fields  (cost=0.00..1556105.80 rows=36205 width=8) (actual time=0.022..0.023 rows=1 loops=1)
         Output: id
         Filter: (nf_fields.local_algs @> '{2.8.16.6}'::text[])
         Rows Removed by Filter: 8
         Buffers: shared hit=1
 Planning:
   Buffers: shared hit=1
 Planning Time: 0.256 ms
 Execution Time: 0.036 ms
(12 rows)
```

so I most likely the timeout was a latency hiccup of the disk

**Comment by roed314** [comment]
It seems to have happened again:
```
psycopg2.errors.QueryCanceled: canceling statement due to statement timeout
 happens while executing b'SELECT "id" FROM "nf_fields" WHERE "local_algs" @> ARRAY[\'5.10.19.5\']::text[] LIMIT 1'
[2021-11-25 13:44:06 UTC] 500 error on URL https://www.lmfdb.org/padicField/5.10.19.5 ()
```
I'm not going to be able to look at this for a few days because of the holidays.

**Comment by assaferan** [comment]
Hi, I can take this one

**Comment by assaferan** [comment]
Actually, I'm not sure I can replicate this. It stopped timing out after the last merge.

**Comment by assaferan** [comment]
After looking at it for a while, it looks like Postgres is choosing to use sequential scan because the statistics for the number of returned rows are off. Look for example at
sage: an_query = b'EXPLAIN (ANALYZE, COSTS, VERBOSE, BUFFERS) SELECT "id" FROM "
....: nf_fields" WHERE "local_algs" @> ARRAY[\'5.10.19.5\']::text[] LIMIT 1'    
sage: cursor.execute(an_query)                                                  
sage: cursor.fetchall()                                                         
[('Limit  (cost=0.00..115.19 rows=1 width=8) (actual time=8019.205..8019.207 rows=1 loops=1)',),
 ('  Output: id',),
 ('  Buffers: shared hit=942762',),
 ('  ->  Seq Scan on public.nf_fields  (cost=0.00..1556618.55 rows=13514 width=8) (actual time=8019.202..8019.203 rows=1 loops=1)',),
 ('        Output: id',),
 ("        Filter: (nf_fields.local_algs @> '{5.10.19.5}'::text[])",),
 ('        Rows Removed by Filter: 16499553',),
 ('        Buffers: shared hit=942762',),
 ('Planning:',),
 ('  Buffers: shared hit=1',),
 ('Planning Time: 0.370 ms',),
 ('Execution Time: 8019.234 ms',)]

Postgres seems to think it will only have to process 13514 rows, but actually 16499554 rows are processed. Maybe we haven't run many such queries to populate the statistics?

**Comment by roed314** [comment]
I believe that the issue is non-uniformness.  Postgres is assuming that the cases satisfying the query are uniformly distributed, while in fact they're highly non-uniform.  So while there are plenty of things to find, they're after all the quadratic fields (for example).  I don't know a good way to solve this; postgres has some [more advanced statistics](https://www.postgresql.org/docs/current/multivariate-statistics-examples.html), but I don't think any of them help in our use case.  Maybe it's worth asking in a database stackexchange or something, since this is an issue we've wrestled with for a long time?

**Comment by assaferan** [comment]
There are already several threads about these issues in stackexchange which I came across when trying to solve this, see e.g.  [this one](https://stackoverflow.com/questions/34537096/postgres-not-using-index-when-index-scan-is-much-better-option), [this one](https://stackoverflow.com/questions/21385555/postgresql-query-very-slow-with-limit-1), [this one](https://stackoverflow.com/questions/8228326/keep-postgresql-from-sometimes-choosing-a-bad-query-plan?noredirect=1&lq=1), or [this one ](https://stackoverflow.com/questions/62258912/postgresql-limit-in-query-makes-it-not-use-index).  The problem is that the offered solutions are not very robust - we don't want to turn off sequential scan in production, to query for more results when we need only one, or to ask them to be sorted. The two last ones should, in principal, be more work, not less... 

This issue also appears on [Postgres](https://www.postgresql.org/message-id/flat/CA%2BU5nMLbXfUT9cWDHJ3tpxjC3bTWqizBKqTwDgzebCB5bAGCgg%40mail.gmail.com).

**Comment by roed314** [comment]
I wonder if we can use MCV multivariate statistics productively.  It has the potential to help with situations where two columns aren't independent.

The other thing I would point out is that indexes should usually contain the columns used for sorting on that table, in the same order at the end.  So, for example, there is an index on `['local_algs']` for number fields that might be better as `['local_algs', 'degree', 'disc_abs', 'disc_sign', 'iso_number']`.  I think that this will help with our normal queries, which almost always include a `LIMIT` and `ORDER BY` clauses: it makes index-only scans possible (which are faster), and means you don't have to get all results with a given value in order to sort (thus sometimes allowing the avoidance of sequential scans).

**Comment by roed314** [comment]
Actually, I think I'm wrong on `local_algs`, since that needs to be a GIN index rather than a btree.  For btree indexes, including the sort columns is a good idea.

**Comment by edgarcosta** [comment]
We should perhaps use the user input to restrict the search further, for example, we can use that the degree is at least the degree of the given field and that the prime is ramified. However, given multiple local fields, this might be useless.

**Comment by edgarcosta** [comment]
Resolved via #5116
